### ISSUE TITLE
Have GitHub actions run `chplspell` spellcheck?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#6731]
At the risk of putting @cassella out of a hobby, I'm thinking that we should consider running chplspell as part of the ~Travis~ GitHub Actions pre-merge checks, assuming we think it's worth the effort to keep our spelling correct (I think it is).  ~I believe I've heard that having Travis run chplspell is not trivial, but I'm not sure what the challenges are.~  I'm opening this issue to (a) see whether there'd be general support for doing this and (b) see what the technical challenges to doing so are.

**Comment by ben-albrecht** [comment]
>(a) see whether there'd be general support for doing this

👍  

> (b) see what the technical challenges to doing so are.

I don't know of any technical challenges. It builds off of a separate virtualenv, so there will be some additional build time.

I do have some concern that maintaining the dictionary at a PR-granularity might be tedious for developers, but I'm open to trying it for a while and finding out how tedious it actually is.

**Comment by cassella** [comment]
>   (b) see what the technical challenges to doing so are.

One challenge that's occurred to me with use is merge conflicts.  The
dictionary files seem prone to conflicting updates.  So PRs that
update the dictionary files and remain open for a long time are likely
to have merge conflicts when it's time to merge.  And manually
resolving conflicts in these files is likely to be tedious.

On the other hand, recreating the dictionary file updates should be
fairly painless.  One approach would be to make the dictionary file
updates in a separate commit.  That way if there's a conflict at merge
time, you can easily rebase to drop the dictionary update commit, and
generate a new one.  That's still kind of a hassle though.

This isn't specific to having Travis run chplspell, but to any
workflow where everyone's keeping the dictionary files up to date with
PR granularity.



Another drawback is that you won't be able to get a clean Travis run
(or smoketest run if it's added to the smoketest) on a branch until
attending to the dictionary files.  I don't know if this would worsen
anyone's workflow.


A third is the risk that chplspell won't work for everyone -- e.g. on
some system configurations, or that the interface isn't as intuitive
as intended.  I think only a small handful of people have tried it to
date.


`chpslpell --report-only` (after the open PRs) currently reports two
issues with no resolution yet:

```test/studies/shootout/submitted/spectralnorm.chpl:41: 'tranpose' not found in dictionary (from token 'tranpose')```

Due to the benchmark having been submitted with the misspelling; and

```compiler/include/resolution.h:148: 'eturn' not found in dictionary (from token 'eturn')```

Due to the use of `\return foo` in doxygen comments.  (scspell treats
`\r` as the C-escaped whitespace character.)  Discussed in PR #5964.
I believe doxygen accepts `@return foo`, but I don't know how to test
everything that consumes those comments.  (Or even identify everything.)



AFAIK, adding chplspell to the Travis run would be about as trivial as
anything else that might be added to Travis.  After clobbering the
venv install dir, `make chplspell-venv` on my machine takes 3.5
seconds.  (Though it's probably got the scspell package cached
somewhere.)  Also on my machine, `chplspell --report-only` takes 13
seconds.

**Comment by cassella** [comment]
> So PRs that update the dictionary files and remain open for a long time are likely to have merge conflicts when it's time to merge.

On the third hand, maybe conflicts won't be very frequent in practice.  A conflict would result from two concurrent PRs
1. adding words near each other alphabetically
2. adding per-file dictionaries for files near each other hierarchically (and alphabetically), conflicting in the file-id mapping file.
3. adding words to per-file dictionaries that are near each other in the dictionary file.  (Due to the construction of the dictionary file, this means files whose per-file dictionaries were created in chronological proximity.)

It's plausible that the dictionary and file id mappings are large enough that conflicts will be rare in practice.  That third case seems the most likely, since any two newly-created per-file dictionaries will be added at the same place in the dictionary file -- just before the natural-language dictionary.  Maybe resolving that sort of conflict won't be too bad since generally all the commits involved would be adding new stanzas.

**Comment by ronawho** [comment]
https://github.com/chapel-lang/chapel/pull/6922 adds chplspell to travis testing. 

I plan to bring this up a discussion item at our next WSM and if nobody objects we can try it out and see how it goes.

**Comment by tzinsky** [comment]
Commits #[6922](https://github.com/chapel-lang/chapel/pull/6922) and #[6977](https://github.com/chapel-lang/chapel/pull/6977) added this capability.  Travis is not being used as the Chapel CI any longer.   We have moved our CI to a Jenkins server, and there are github actions currently in place.

**Comment by bradcray** [comment]
@tzinsky : I'm confused b/c I don't believe we've ever had this capability enabled (so whatever happened in those commits didn't cause it to happen, I don't think?)  I understand that we don't use Travis anymore, but also don't believe we're doing any automated spellchecking as part of our GitHub actions (Specifically, I'm sure Paul is still fixing spellchecking errors that I've introduced and that the CI hasn't caught me on).  So I would've either updated this issue's title and OP to be "GitHub actions run spellcheck?" or open a new one with that title and description.

**Comment by lydia-duncan** [comment]
#6922 was never merged and #6977 impacts `chplspell`'s functionality but not whether it turns up in our CI testing

**Comment by bradcray** [comment]
Thanks Lydia.  So would you agree that we should either reopen this, editing to replace Travis with GitHub actions or create a new issue?  (I'd probably do the former to not lose the previous history, unless it's all Travis-related, but I don't think it was).

**Comment by lydia-duncan** [comment]
Yeah, I agree that we should reopen it and edit to replace Travis with Github Actions
